R/utilities-simulation.R

```diff
@@ -749,3 +749,230 @@ getSimulationTree <- function(simulationOrFilePath, quantityType = "Quantity") {
 
   return(pathEnumList)
 }
+
+#' Get the steady-state values of species and state variable parameters.
+#'
+#' @details The steady-state is considered to be the last values of the molecules
+#' amounts and state variable parameters in the simulation with sufficiently long
+#' simulation time, i.e., where the rates of
+#'   the processes do not (significantly) change. The steady-state is NOT analytically
+#'   calculated or estimated in any other way than simulating for the given time.
+#'
+#'
+#' @param steadyStateTime Simulation time (minutes). Must be long enough for
+#'   system to reach a steady-state. 1000 by default. Either a single value (will
+#'   be applied for all simulations), or a vector of values specific for each
+#'   simulation. In latter case, must have equal size as `simulations`.
+#' @param quantitiesPaths List of quantity paths (molecules and/or parameters)
+#'   for which the steady-state will be simulated. If `NULL` (default), all
+#'   molecules and state variable parameters are considered. The same list is
+#'   applied for all simulations.
+#' @param simulations `Simulation` object or a list of `Simulation` objects
+#' @param ignoreIfFormula If `TRUE` (default), species and parameters with
+#'   initial values defined by a formula are not included.
+#' @param stopIfNotFound Boolean. If `TRUE` (default), an error is thrown when
+#'   results for certain species were not generated. This may happen when due to
+#'   numerical problems some values cannot be calculated, though the whole
+#'   simulation converges. Setting this argument to `FALSE` allows to ignore
+#'   such errors. Check the outputs for empty values when using this option.
+#' @param lowerThreshold Numerical value (in default unit of the output).
+#' Any steady-state values below this value are considered as numerical noise
+#' and replaced by 0. If `lowerThreshold` is `NULL`, no cut-off is applied.
+#' Default value is 1e-15.
+#' @param simulationRunOptions Optional instance of a `SimulationRunOptions`
+#'  used during the simulation run.
+#'
+#' @return A named list, where the names are the IDs of the simulations and the
+#'   entries are lists containing `paths` and their `values` at the end of the
+#'   simulation.
+#' @import ospsuite.utils
+#' @export
+#' @examples
+#' simPath <- system.file("extdata", "Aciclovir.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' steadyState <- getSteadyState(simulations = sim)
+#' # Set initial values for steady-state simulations
+#' setQuantityValuesByPath(
+#'   quantityPaths = steadyState[[sim$id]]$paths,
+#'   values = steadyState[[sim$id]]$values, simulation = sim
+#' )
+getSteadyState <- function(quantitiesPaths = NULL,
+                           simulations,
+                           steadyStateTime = 1000,
+                           ignoreIfFormula = TRUE,
+                           stopIfNotFound = TRUE,
+                           lowerThreshold = 1e-15,
```

Comment:
In the Ospsuite.Core we calculate a Comparison Threshold for each state variable and each observer dynamically, based on the AbsTol and ScaleFactor for the state variables and on the monitor formula for observers; also having special handling for Fractions.
Values below that comparison tolerance are considered numerical noise.
I wonder if we should export this property to R and use as a default for each variable if not passed by user

Reply:
I am not sure this works properly... This is what I wanted to avoid:

![image](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/assets/25903040/1291ceb3-08a8-46c1-81c1-27b1e9ac72c1)

Amount of an orally administered drug in stomach.

Reply:
Hmm, it's theoretically done in SimModel:
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/7092dd785819f9904abbc1f40cc33c8e4be05a30/src/OSPSuite.SimModelNative/src/DESolver.cpp#L350-L364

However when looking into the implementation: only the values in `[-Threshold .. 0]` are set to zero, but **not the values in** `[0 .. Threshold]`
https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/7092dd785819f9904abbc1f40cc33c8e4be05a30/src/OSPSuite.SimModelNative/src/SimulationTask.cpp#L186C1-L193C1

I assume originally the values in [-Threshold .. Threshold] were set to zero, but we got some kind of problems with this (but I don't remember which exactly..). Will create an issue for this in SimModel

Reply:
s. https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/issues/162